File: src/lib/watermark/kodak-slide.ts
import { normalizeModel } from "@/lib/exif/brand";
import type { FrameParams, TemplateConfig } from "@/stores/types";
import {
  formatTakenAt,
  getPreviewFontFamily,
  sanitizeCustomLines,
} from "./classic-bottom";
import {
  dprFor,
  fitPhoto,
  getCanvasRatio,
  type ShowcasePhoto,
} from "./renderer-utils";

const SLIDE_PAPER = "#f5f3ed";
const SLIDE_FRAME = "#fafaf6";
const SLIDE_FRAME_SHADOW = "rgba(0,0,0,0.16)";
const SLIDE_INK = "#1a1a1a";
const KODAK_RED = "#cb1f27";
const KODAK_YELLOW = "#ffce00";

/**
 * Kodak slide / Kodachrome 35mm transparency look. Cream backdrop with a
 * slightly off-white slide-mount frame around the photo; the bottom frame
 * carries a red "KODACHROME" stripe and date stamp typical of mounted slides.
 */
export function drawKodakSlidePreview(
  canvas: HTMLCanvasElement,
  image: HTMLImageElement,
  _logo: HTMLImageElement | null,
  photo: ShowcasePhoto,
  frameParams: FrameParams,
  config: TemplateConfig,
) {
  const ctx = canvas.getContext("2d");
  if (!ctx) return;

  const dpr = dprFor(canvas);
  const baseWidth = 900;
  const ratio = getCanvasRatio(frameParams.canvasRatio, frameParams.canvasOrientation);
  const canvasW = baseWidth;
  const canvasH = canvasW / ratio;
  const watermarkActive = config.showWatermark ?? true;

  // Slide mount: thicker bottom edge to host the Kodachrome stripe.
  const sideFrame = canvasW * 0.05;
  const topFrame = canvasH * 0.07;
  const bottomFrame = watermarkActive ? Math.max(72, canvasH * 0.16) : topFrame;

  const slideX = canvasW * 0.06;
  const slideY = canvasH * 0.05;
  const slideW = canvasW - slideX * 2;
  const slideH = canvasH - slideY * 2;
  const photoArea = {
    x: slideX + sideFrame,
    y: slideY + topFrame,
    w: slideW - sideFrame * 2,
    h: slideH - topFrame - bottomFrame,
  };
  const placed = fitPhoto(photoArea, photo.width, photo.height);

  canvas.width = Math.round(canvasW * dpr);
  canvas.height = Math.round(canvasH * dpr);
  canvas.style.aspectRatio = `${canvasW} / ${canvasH}`;
  ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
  ctx.clearRect(0, 0, canvasW, canvasH);

  ctx.fillStyle = SLIDE_PAPER;
  ctx.fillRect(0, 0, canvasW, canvasH);

  // Slide mount with a subtle drop shadow.
  ctx.save();
  ctx.shadowColor = SLIDE_FRAME_SHADOW;
  ctx.shadowBlur = 18;
  ctx.shadowOffsetY = 4;
  ctx.fillStyle = SLIDE_FRAME;
  ctx.fillRect(slideX, slideY, slideW, slideH);
  ctx.restore();

  ctx.drawImage(image, placed.x, placed.y, placed.w, placed.h);

  if (!watermarkActive) return;

  // Photo cutout border (1px black) — emulates slide aperture cut.
  ctx.strokeStyle = "#1a1a1a";
  ctx.lineWidth = 1;
  ctx.strokeRect(placed.x - 0.5, placed.y - 0.5, placed.w + 1, placed.h + 1);

  const fontFamily = getPreviewFontFamily(frameParams.fontFamily);
  const stripeTop = placed.y + placed.h + canvasH * 0.025;
  const stripeH = Math.max(22, canvasH * 0.04);
  const stripeW = slideW - sideFrame * 2;

  // KODACHROME red stripe.
  ctx.fillStyle = KODAK_RED;
  ctx.fillRect(placed.x, stripeTop, stripeW, stripeH);

  // Inset yellow rule on top edge.
  ctx.fillStyle = KODAK_YELLOW;
  ctx.fillRect(placed.x, stripeTop, stripeW, 2);

  // KODACHROME wordmark left + date right.
  ctx.fillStyle = "#ffffff";
  ctx.textBaseline = "middle";
  ctx.textAlign = "left";
  const wordSize = Math.max(11, frameParams.fontSize * 1.1);
  ctx.font = `800 ${wordSize}px ${fontFamily}`;
  ctx.fillText("KODACHROME 64", placed.x + stripeH * 0.5, stripeTop + stripeH / 2);

  const date = config.showDate ? formatTakenAt(photo.exif.takenAt, config.dateFormat) : "";
  if (date) {
    ctx.textAlign = "right";
    ctx.font = `600 ${wordSize}px ${fontFamily}`;
    ctx.fillText(date, placed.x + stripeW - stripeH * 0.5, stripeTop + stripeH / 2);
  }

  // Camera under stripe in slide-mount black ink.
  const camera = config.showCamera
    ? normalizeModel(photo.exif.camera.make, photo.exif.camera.model)
    : "";
  const customLines = sanitizeCustomLines(config.customLines);
  const subline = camera || customLines[0] || "";
  if (subline) {
    ctx.fillStyle = SLIDE_INK;
    ctx.textAlign = "center";
    ctx.font = `500 ${Math.max(10, frameParams.fontSize)}px ${fontFamily}`;
    ctx.fillText(subline, placed.x + stripeW / 2, stripeTop + stripeH + Math.max(14, frameParams.fontSize));
  }
}

File: src-tauri/src/render/templates/kodak_slide.rs
use image::{imageops::overlay, DynamicImage, Rgba, RgbaImage};

use crate::render::classic_bottom::{
    compute_canvas_size, fill_rect, fit_photo_in_area, format_camera, format_date, resize_photo,
    ExportExif, ExportFrameParams, ExportTemplateConfig,
};
use crate::render::text::TextRenderer;

const SLIDE_PAPER: Rgba<u8> = Rgba([245, 243, 237, 255]);
const SLIDE_FRAME: Rgba<u8> = Rgba([250, 250, 246, 255]);
const SLIDE_SHADOW: Rgba<u8> = Rgba([0, 0, 0, 41]);
const SLIDE_INK: Rgba<u8> = Rgba([26, 26, 26, 255]);
const KODAK_RED: Rgba<u8> = Rgba([203, 31, 39, 255]);
const KODAK_YELLOW: Rgba<u8> = Rgba([255, 206, 0, 255]);
const KODAK_WHITE: Rgba<u8> = Rgba([255, 255, 255, 255]);

pub(crate) fn compose_kodak_slide(
    source: DynamicImage,
    frame: &ExportFrameParams,
    exif: &ExportExif,
    config: &ExportTemplateConfig,
    text: Option<&TextRenderer>,
) -> RgbaImage {
    let src_w = source.width();
    let src_h = source.height();
    let (canvas_w, canvas_h) = compute_canvas_size(src_w, src_h, frame);
    let rs = src_w as f32 / 900.0;

    let side_frame = (canvas_w as f32 * 0.05).round() as u32;
    let top_frame = (canvas_h as f32 * 0.07).round() as u32;
    let bottom_frame = if config.show_watermark {
        ((canvas_h as f32 * 0.16).round() as u32).max((72.0 * rs) as u32)
    } else {
        top_frame
    };

    let slide_x = (canvas_w as f32 * 0.06).round() as u32;
    let slide_y = (canvas_h as f32 * 0.05).round() as u32;
    let slide_w = canvas_w.saturating_sub(slide_x * 2);
    let slide_h = canvas_h.saturating_sub(slide_y * 2);

    let area_x = slide_x + side_frame;
    let area_y = slide_y + top_frame;
    let area_w = slide_w.saturating_sub(side_frame * 2);
    let area_h = slide_h.saturating_sub(top_frame + bottom_frame);
    let (pw, ph, ox, oy) = fit_photo_in_area(area_w, area_h, src_w, src_h);
    let placed_x = area_x + ox;
    let placed_y = area_y + oy;

    let mut canvas = RgbaImage::from_pixel(canvas_w, canvas_h, SLIDE_PAPER);

    let shadow_offset = (4.0 * rs).round() as i64;
    fill_rect(&mut canvas, slide_x as i64 + shadow_offset, slide_y as i64 + shadow_offset, slide_w, slide_h, SLIDE_SHADOW);
    fill_rect(&mut canvas, slide_x as i64, slide_y as i64, slide_w, slide_h, SLIDE_FRAME);

    let photo = resize_photo(source, pw, ph);
    overlay(&mut canvas, &photo, placed_x as i64, placed_y as i64);

    if !config.show_watermark {
        return canvas;
    }

    // 1px black aperture cutout border around the photo (slide-mount feel).
    let stroke = (1.0 * rs).round() as u32;
    let sw = stroke.max(1);
    fill_rect(&mut canvas, placed_x as i64, placed_y as i64 - sw as i64, pw, sw, SLIDE_INK);
    fill_rect(&mut canvas, placed_x as i64, (placed_y + ph) as i64, pw, sw, SLIDE_INK);
    fill_rect(&mut canvas, placed_x as i64 - sw as i64, placed_y as i64 - sw as i64, sw, ph + sw * 2, SLIDE_INK);
    fill_rect(&mut canvas, (placed_x + pw) as i64, placed_y as i64 - sw as i64, sw, ph + sw * 2, SLIDE_INK);

    let Some(rend) = text else { return canvas; };

    let stripe_top = placed_y as f32 + ph as f32 + (canvas_h as f32 * 0.025).round();
    let stripe_h = ((canvas_h as f32 * 0.04).round() as u32).max((22.0 * rs) as u32);
    let stripe_x = area_x;
    let stripe_w = area_w;

    fill_rect(&mut canvas, stripe_x as i64, stripe_top as i64, stripe_w, stripe_h, KODAK_RED);
    fill_rect(&mut canvas, stripe_x as i64, stripe_top as i64, stripe_w, 2, KODAK_YELLOW);

    let word_pt = (frame.font_size as f32 * 1.1).max(11.0 * rs);
    let text_y = stripe_top + stripe_h as f32 / 2.0 - rend.line_height(word_pt, true) / 2.0;
    let pad = stripe_h as f32 * 0.5;

    rend.draw(&mut canvas, "KODACHROME 64", stripe_x as f32 + pad, text_y, word_pt, true, KODAK_WHITE);

    if config.show_date {
        let date = format_date(&exif.taken_at, &config.date_format);
        if !date.is_empty() {
            let dw = rend.measure(&date, word_pt, false);
            let x = stripe_x as f32 + stripe_w as f32 - pad - dw;
            rend.draw(&mut canvas, &date, x, text_y, word_pt, false, KODAK_WHITE);
        }
    }

    let camera = if config.show_camera { format_camera(exif) } else { String::new() };
    let subline = if !camera.is_empty() {
        camera
    } else {
        config.custom_lines.first().cloned().unwrap_or_default().trim().to_string()
    };
    if !subline.is_empty() {
        let sub_pt = (frame.font_size as f32).max(10.0 * rs);
        let sw = rend.measure(&subline, sub_pt, false);
        let sx = stripe_x as f32 + stripe_w as f32 / 2.0 - sw / 2.0;
        let sy = stripe_top + stripe_h as f32 + (14.0 * rs).max(frame.font_size as f32);
        rend.draw(&mut canvas, &subline, sx, sy, sub_pt, false, SLIDE_INK);
    }

    canvas
}
